SSA/Projects/InstCombine/HackersDelight.lean

```diff
@@ -164,16 +169,139 @@ theorem neq_iff_sub_or_sub :
     x ≠ y ↔ (x - y ||| y - x).getMsb 0 := by
   sorry
 
+theorem neq_iff_neg_sub_abs :
+    x ≠ y ↔ (-(x - y).abs).getMsb 0 := by
+  sorry
+
 theorem lt_iff_sub_xor_xor_and_sub_xor :
-    x < y ↔ ((x - y) ^^^ ((x ^^^ y) &&& ((x - y) ^^^ x))).getMsb 0 := by
+    (x <ₛ y) ↔ ((x - y) ^^^ ((x ^^^ y) &&& ((x - y) ^^^ x))).getMsb 0 := by
   sorry
 
 theorem lt_iff_and_not_or_not_xor_and_sub :
-    x < y ↔ ((x &&& ~~~ y) ||| (~~~ (x ^^^ y) &&& (x - y))).getMsb 0 := by
+    (x <ₛ y) ↔ ((x &&& ~~~ y) ||| (~~~ (x ^^^ y) &&& (x - y))).getMsb 0 := by
   sorry
 
 theorem le_iff_or_not_and_xor_or_not_sub :
-    x ≤ y ↔ ((x ||| ~~~ y) &&& ((x ^^^ y) ||| ~~~ (y - x))).getMsb 0 := by
+    (x ≤ₛ y) ↔ ((x ||| ~~~ y) &&& ((x ^^^ y) ||| ~~~ (y - x))).getMsb 0 := by
+  sorry
+
+theorem ult_iff_not_and_or_not_xor_and_sub :
+    (x <ᵤ y) ↔ ((~~~ x &&& y) ||| (~~~ (x ^^^ y) &&& (x - y))).getMsb 0 := by
+  sorry
+
+theorem ule_iff_not_or_and_xor_or_not_sub :
+    (x ≤ᵤ y) ↔ ((~~~ x ||| y) &&& ((x ^^^ y) ||| ~~~ (y - x))).getMsb 0 := by
+  sorry
+
+theorem eq_zero_iff_abs_sub :
+    x = 0 ↔ (x.abs - 1).getMsb 0 := by
+  sorry
+
+theorem eq_zero_iff_not_or_sub :
+    x = 0 ↔ (~~~ (x ||| -x)).getMsb 0 := by
+  sorry
+
+theorem eq_zero_iff_not_and_sub :
+    x = 0 ↔ (~~~ x &&& x - 1).getMsb 0 := by
+  sorry
+
+theorem neq_zero_iff_or_neg :
+    x ≠ 0 ↔ (x ||| -x).getMsb 0 := by
+  sorry
+
+theorem neq_zero_iff_neg_abs :
+    x ≠ 0 ↔ (-(x.abs)).getMsb 0 := by
+  sorry
+
+theorem slt_zero_iff :
+    (x <ₛ 0) ↔ x.getMsb 0 := by
+  sorry
+
+theorem sle_zero_iff_or_sub_one :
+    (x ≤ₛ 0) ↔ (x ||| (x - 1)).getMsb 0 := by
+  sorry
+
+theorem sle_zero_iff_or_not_sub :
+    (x ≤ₛ 0) ↔ (x ||| ~~~ (-x)).getMsb 0 := by
+  sorry
+
+theorem zero_slt_iff_neg_and_not :
+    (0 <ₛ x) ↔ (-x &&& ~~~ x).getMsb 0 := by
+  sorry
+
+theorem zero_sle_iff_neg_and_not :
+    (0 ≤ₛ x) ↔ (~~~ x).getMsb 0 := by
+  sorry
+
+theorem slt_iff_add_two_pow_ult_add_two_pow :
+    (x <ₛ y) ↔ (x + 2 ^ (w - 1) <ₛ y + 2 ^ (w - 1)) := by
+  sorry
+
+theorem ult_iff_sub_two_pow_ult_sub_two_pow :
+    (x <ᵤ y) ↔ (x - 2 ^ (w - 1) <ₛ y - 2 ^ (w - 1)) := by
+  sorry
+
+theorem slt_iff_not_add_two_pow_ule_add_two_pow :
+    (x <ₛ y) ↔ ¬ ((y + 2 ^ (w - 1)) ≤ᵤ (x + 2 ^ (w - 1))) := by
+  sorry
+
+theorem sle_iff_add_two_pow_ule_add_two_pow :
+    (x ≤ₛ y) ↔ (x + 2 ^ (w - 1)) ≤ᵤ (y + 2 ^ (w - 1)) := by
+  sorry
+
+theorem ult_iff_not_ule :
+    (x <ᵤ y) ↔ ¬ (y ≤ᵤ x) := by
+  sorry
+
+theorem eq_iff_adc_not_add :
+    x = y ↔ (BitVec.carry w (x) (~~~ y + 1)) false := by
+  sorry
+
+theorem neq_iff_adc_not :
+    x ≠ y ↔ (BitVec.carry w (x) (~~~ y)) false := by
+  sorry
+
+theorem slt_iff_not_adc_add_sub_neg_add_sub :
+    (x <ₛ y) ↔  ¬ (BitVec.carry w (x + 2 ^ (w - 1)) (~~~ (y + 2 ^ (w - 1)) + 1)) false := by
+  sorry
+
+instance : HXor Bool Bool Bool where
```

Comment:
Is this a hack? I think it might be better to do a local instance
https://lean-lang.org/theorem_proving_in_lean4/type_classes.html#local-instances

Reply:
So that this instance is only used inside this namespace.